File: services/c-tcp-broker/send_client.c
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/shm.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <linux/sockios.h>
#include <netdb.h>
#include <stdlib.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>

#include "../lib/message.h"
#include "../lib/network.h"
#include "../lib/timer.h"


#define MAX_COUNT 100000
#define WS_1 1

int recv_msg(int fd){
  struct ack_message msg_ack;

  net_recv_ack(fd, &msg_ack, SEND_ACK);

  return msg_ack.hdr.ws;
}


int send_msg (int fd,
	      uint32_t n,
	      uint32_t saddr,
	      uint32_t daddr,
	      void *payload){
  struct message smsg;
  uint64_t log_tsc;
  msg_fill(&smsg,SEND_MSG, n, saddr, daddr, payload,sizeof(payload));
  log_tsc = getclock();
  msg_assign_time_stamp(&smsg, log_tsc, SENDER_SEND);
  net_send_msg(fd, &smsg);

  return 0;
}

void send_msgs (int count, uint32_t win_size, char *host, int port_num){
  int fd = socket(AF_INET, SOCK_STREAM, 0);

  if (fd < 0) {
    perror("socket\n");
    return;
  }
  struct addrinfo hints, *res;
  struct in_addr in_addr;
  int err;
  struct sockaddr_in addr;
  
  memset(&hints, 0, sizeof(hints));
  hints.ai_socktype = SOCK_STREAM;
  hints.ai_family = AF_INET;
  
  if ((err = getaddrinfo(host, NULL, &hints, &res)) != 0) {
    printf("error %d\n", err);
    return;
  }
  in_addr.s_addr= ((struct sockaddr_in *)(res->ai_addr))->sin_addr.s_addr;

  memset(&addr, 0, sizeof(addr));
  addr.sin_family = AF_INET;
  addr.sin_port = htons(port_num);
  memcpy(&addr.sin_addr, &in_addr, sizeof(in_addr));

  //struct hostent *hp;
  //if ((hp = gethostbyname(host)) == NULL) {
  //  fprintf(stderr, "gethost error %s\n", host);
  //  close(fd);
  //  return;
  //}
  //
  //struct sockaddr_in addr;
  //memset(&addr, 0, sizeof(addr));
  //addr.sin_family = AF_INET;
  //addr.sin_port = htons(port_num);
  //memcpy(&addr.sin_addr, hp->h_addr, hp->h_length);
  //
  //print_hostname(hp);
  //print_ipaddr_v4(hp);
  

  while (1) {
    if (connect(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
      sleep(1);
      printf("connect\n");
      continue;
    } else {
      break;
    }
  }

  int on=1;
  int ret;
  ret = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &on, sizeof(on));

  //int fd = net_connect(host, port_num);

  //start_send_messages
  char payload[MSG_PAYLOAD_LEN] = "Hello";
  uint32_t saddr = 100;
  uint32_t daddr = 200;
  uint32_t ws;
  uint32_t next_ws;

  int send_count = 0;

  net_hello_req(fd, saddr, daddr);

  ws = WS_1;

  while(send_count + ws < count){
    for (uint32_t i = ws; i > 0; i--) {
      send_msg(fd, i, saddr, daddr, payload);
    }
    next_ws = recv_msg(fd);
    send_count += ws;
    //ws = next_ws;
    ws = win_size;
  }
  for (uint32_t i = (count - send_count); i > 0; i--){
    send_msg(fd, i, saddr, daddr, payload);
  }
  recv_msg(fd);

  //end_sendmessages

  if (close(fd) == -1) {
    printf("%d\n", errno);
  }
}

int main (int argc, char *argv[])
{
  int count = 1;
  uint32_t win_size = 1;
  char host_name[80] = "localhost";
  int port_num = 8000;

  if(argc > 1){
    count = atoi(argv[1]);
  }else{
    printf("%s argument error count\n", __FILE__);
    return 0;
  }

  if(argc > 2){
    win_size = atoi(argv[2]);
  }else{
    printf("%s argument error win_size\n", __FILE__);
    return 0;
  }

  if(argc > 3 && strlen(argv[3]) < 81){
    strncpy(host_name, argv[3], strlen(argv[3])+1);
  }else{
    printf("%s argument error host_name\n", __FILE__);
    return 0;
  }

  if(argc > 4){
    port_num = atoi(argv[4]);
  }else{
    printf("%s argument error port_num\n", __FILE__);
    return 0;
  }

  send_msgs(count, win_size, host_name, port_num);
  return 0;
}

File: services/ruby-tcp-broker/send_client.rb
require "../lib/message.rb"
require "../lib/network.rb"
require "../lib/timer.rb"
require "socket"

WS_1 = 1
$time_count = 0

def recv_msg(s)
  msg = Ack_Message.new
  net_recv_ack(s, msg, SEND_ACK)
  ws = msg.ws
end

def send_msg(s, fragments, saddr, daddr, payload)
  msg = Message.new
  msg = msg_fill(msg, SEND_MSG, fragments, saddr, daddr, payload)
  time = getclock()
  msg_assign_time_stamp(msg, time, SENDER_SEND)
  net_send_msg(s, msg)
end

def send_msgs(host, count, data_size, win_size, port_num)
  s = TCPSocket.open(host, port_num)
  s.setsockopt(Socket::IPPROTO_TCP,Socket::TCP_NODELAY,true)

  #start_send_messages
  payload = "Hello"
  saddr = 100
  daddr = 200
  next_ws = WS_1

  send_count = 0

  net_hello_req(s, saddr, daddr)

  ws = WS_1

  while (send_count + ws) < count do
    ws.downto(1) do |i|
      send_msg(s, i, saddr, daddr, payload)
    end
    next_ws = recv_msg(s)
    send_count += ws
    #ws = next_ws
    ws = win_size
  end

  left_count = count - send_count
  left_count.downto(1) do |i|
    send_msg(s, i, saddr, daddr, payload)
  end
  recv_msg(s)
#end_send_msgs

  s.close
end

def main()
  if ARGV.size > 0
    count = ARGV[0].to_i
  else
    file = File.basename(__FILE__)
    STDERR.printf("%s argument error count\n", file)
    exit
  end

  if ARGV.size > 1
    data_size = ARGV[1].to_i
  else
    file = File.basename(__FILE__)
    STDERR.printf("%s argument error data_size\n", file)
    exit
  end

  if ARGV.size > 2
    win_size = ARGV[2].to_i
  else
    file = File.basename(__FILE__)
    STDERR.printf("%s argument error window_size\n", file)
    exit
  end

  if ARGV.size > 3
    port_num = ARGV[3].to_i
  else
    file = File.basename(__FILE__)
    STDERR.printf("%s argument error port_num\n", file)
    exit
  end

  if (count < win_size)
    puts"count < window_size"
    exit
  end

  send_msgs("localhost", count, data_size, win_size, port_num)
end

main

